src/protocol/check_zero.rs

```diff
@@ -0,0 +1,149 @@
+};
+
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've multiplied their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct DValue<F> {
+    d: F,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    MultiplyWithR,
+    RevealR,
+}
+
+impl crate::protocol::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::MultiplyWithR => "multiply_with_r",
+            Self::RevealR => "reveal_r",
+        }
+    }
+}
+
+/// A very simple protocol to check if a replicated secret sharing is in fact a sharing of zero
+///
+/// NOTE: this protocol leaks information about `v` the helpers. Please only use this in cases where
+/// this type of information leakage is acceptable, such as where `v` is the product of a secret value
+/// and a random, unknown value.
+///
+/// This is an implementation of PROTOCOL 3.7 from the paper:
+/// Fast Large-Scale Honest-Majority MPC for Malicious Adversaries
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// The parties start out holding a replicated secret sharing of a value `v`, which they would like to check for equality to zero.
+/// First, the parties generate a secret sharing of a random value `r`, which is not known to any of them.
+/// Next, the parties compute a secret sharing of `r * v` using a multiplication protocol that is secure up to an additive attack.
+/// Then, the parties reveal (sometimes called "open") the secret sharing of `r * v`
+/// If `v` was a secret sharing of zero, this revealed value will also be zero.
+/// On the other hand, if `v` was NOT a secret sharing of zero, then there are two possibilities:
+/// 1.) If the randomly chosen value `r` just so happenned to be zero, then the revealed value will be zero.
+/// This will happen with probability `1/|F|` (where `|F|` denotes the cardinality of the field)
+/// 2.) If the randomly chosen value `r` is any other value in the field aside from zero, then the revealed value will NOT be zero.
+///
+/// Clearly, the accuracy of this protocol is highly dependent on the field that is used.
+/// In a large field, like the integers modulo 2^31 - 1, the odds of `check_zero` returning "true"
+/// when `v` is NOT actually a sharing of zero is extrmely small; it is less than one two billion odds.
+/// In a silly field, like our test field of the integers modulo 31, the odds are very good. It'll incorrectly return "true"
+/// about 3.2% of the time.
+///
+/// ## Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+#[allow(dead_code)]
+pub async fn check_zero<F: Field, N: Network>(
+    ctx: ProtocolContext<'_, N>,
+    record_id: RecordId,
+    v: Replicated<F>,
+) -> Result<bool, BoxError> {
+    let prss = &ctx.prss();
+    let (left, right) = prss.generate_fields(record_id.into());
```

Comment:
I see this a lot.  Should we make the functions on the PRSS take `impl Into<u128>` or `RecordId`?

Reply:
That's a good idea. I think we should make them also accept RecordId directly. I'll put that in a follow-up diff to keep things logically separated.

Reply:
Note that taking RecordId means that you can't feed it a bare number, which could be annoying for testing purposes.  Making the function generic avoids that issue, but also ensures that when you compile the real system you only pay for what you need.

Reply:
Can we just implement two versions of this function which accept different arguments?

Reply:
Rust doesn't do overloading like C++, so you would need two names.